### PR TITLE
Readable error messages from Ad parsing

### PR DESCRIPTION
**Pull request by keileg**
This PR attempts to provide more informative messages when errors (typically dimension mismatches) are raised during parsing of Ad equations. The format of the error messages, including which information to provide, should be improved, but for that we need more experience.

Also, the PR contains some minor maintenance of the Ad framework.

### CODE REVIEW

File: src/porepy/numerics/ad/grid_operators.py [hunk 1]
```diff
@@ -294,32 +298,52 @@ def __init__(
...
         # Vertical stacking of the projections
-        self.primary_to_mortar_int = bmat([[m] for m in primary_to_mortar_int])
-        self.primary_to_mortar_avg = bmat([[m] for m in primary_to_mortar_avg])
-        self.secondary_to_mortar_int = bmat([[m] for m in secondary_to_mortar_int])
-        self.secondary_to_mortar_avg = bmat([[m] for m in secondary_to_mortar_avg])
+        self.primary_to_mortar_int = bmat(
+            [[m] for m in primary_to_mortar_int], name="PrimaryToMortarInt"
+        )
+        self.primary_to_mortar_avg = bmat(
+            [[m] for m in primary_to_mortar_avg], name="PrimarytoMortarAvg"
```
Comment (IvarStefansson) [inline src/porepy/numerics/ad/grid_operators.py:325]:
```suggestion
            [[m] for m in primary_to_mortar_avg], name="PrimaryToMortarAvg"
```

File: src/porepy/numerics/ad/grid_operators.py [hunk 2]
```diff
@@ -412,6 +438,12 @@ def __repr__(self) -> str:
         )
         return s
 
+    def __str__(self) -> str:
```
Comment (IvarStefansson) [inline src/porepy/numerics/ad/grid_operators.py:441]: We are introducing a lot of very similar __str__ methods. Would it be possible to do this by inheritance from Operator, using a method that converts the class type to string?

**Review by IvarStefansson — COMMENTED** [review]
Looks good to me. This will be enormously useful!

File: src/porepy/numerics/ad/grid_operators.py (hunk 2 above)
Comment (keileg) [reply src/porepy/numerics/ad/grid_operators.py:441]: There are subtle, but important, nuances between the different string methods. For instance, the close relation to the class name applies to many, but far from all, string methods. I think it is better to spell it out for each class; at least as long as the whole Ad concept is under heavy development.

**Pull request merged**